### ISSUE TITLE
Extend enforced checkstyle rules

### ISSUE THREAD

**Issue by HannesWell** [jgrapht/jgrapht#1052]
```
 * JGraphT version: 1.5.1
 * Java version all/platform: all 
```

**Issue**
In the process of reviewing #1020 the discussion touched the topic of adding check-style rules to to enforce the java naming convention for variables.

I want to take the opportunity to additionally suggest to enforce the use of braces for all `if`-, `else`-, `for`-, `while`- and `do`-statements. Omitting braces for block-control structures is considered bad style because it can, in conjunction with missing formatting, cause false understanding of the code and therefore reduces the maintainability of the code and is error prone. I spotted missing braces quite often in JGraphT.

Besides that I suggest also to enforce `@Override` annotations to overriding methods.

Maybe you or others have more checkstyle rules you would consider useful for JGraphT? If so, we can collect and discuss them in this issue.

If we agree to extend the applied rules I offer to fix the current violations.

**Comment by jkinable** [comment]
I'm not a fan of enforcing braces everywhere. It makes the code a lot more verbose, especially for common small single line if statements. Our code formatter which we run during every release makes the code nice and pretty and takes care of potential 'readability' issues.

Enforcing @Override annotations is however a good idea.

**Comment by jsichi** [comment]
Being tolerant on checkin for things we can autocorrect per-release via automated formatting allows us to be friendlier to new contributors while preserving a uniform style over time.  So let's not add the braces rule.

**Comment by HannesWell** [comment]
@jkinable your right that the you often add another line and some characters that is technically not necessary. But on the other hand sometimes a little white space can improve the structure of the code and its readability. But that's my personal preference and probably depends on what one is used to.
In conjunction with @jsichi's argument I agree to discard the idea of a braces rule.

As far as I know the Eclipse-formatter cannot add braces. But it could be done with a clean-up profile as pre-release action. A clean-up can also include the formatting (and may other actions to automatically clean up the code). But unfortunately I could not found an equivalent _CleanUpApplication_. So a clean-up would need to be performed from the Eclipse-IDE.
And additionally there is a long-existing bug in the add-braces clean-up that comments that are after e.g. an if-statement without braces are accidentally removed. So after applying the clean up the result must be checked if any-comments are removed.
I have to admit that causes more manual work than before.

Furthermore I checked the checkstyle rule `MissingOverride` and noticed that it is already present in JGraphT's checkstyle rule set. But unfortunately it only "Verifies that the @Override annotation is present when the @inheritDoc javadoc tag is present." (see https://checkstyle.sourceforge.io/config_annotation.html#MissingOverride_Description)...
On the other hand, the Eclipse Clean-Up can add the `@Override` annotations properly, but not as command line application...

So what is only left is the enforcement of the java-naming convention John talked about in the mentioned PR. Do you want to make it happen?
The default java rules that are mostly used are
- camel-case with initial lower-case first letter for non static-final variables
- Upper-case only with underscores for static-final variables (constants)
- Camel-case with initial upper-case letter for classes and interfaces
This is an overview of the checkstyle-rules associated to naming conventions:
https://checkstyle.sourceforge.io/config_naming.html

**Comment by jsichi** [comment]
Yes, please go ahead just the naming convention enforcement for now.  I'm guessing there may be a ton of violations to clean up, especially in tests.

I guess the lack of auto- insertion for braces is something we lost way back when we had to move from Jalopy to Eclipse, too bad.  Jalopy was really good but it couldn't keep up with the Java times.

**Comment by HannesWell** [comment]
Alright.

I've looked into the naming convention rules in deeper detail and suggest the following rules:

-CatchParameterName                     
-ClassTypeParameterName                 
-InterfaceTypeParameterName             
-IllegalIdentifierName                  
-LambdaParameterName                    
-LocalFinalVariableName                 
-LocalVariableName                      
-MemberName                             
-MethodName                             
-MethodTypeParameterName                
-PackageName                            
-ParameterName                          
-PatternVariableName                    
-RecordComponentName                    
-RecordTypeParameterName                
-StaticVariableName                     
-TypeName                                                    

These are all of the rules listed here (https://checkstyle.sourceforge.io/config_naming.html) except for the `AbbreviationAsWordInName` and `AbstractClassName`. The latter would require that abstract class' names have the prefix `Abstract` which would break JGraphT's naming schema that often uses `Base` instead.

Some rules affect some features that are not yet present in Java-8. But to make JGraphT future-prove I suggest to enforce the rules, even it has almost no effect in the moment.

Regarding the *TypeParameterName I'm not sure if you want them. Those enforce that the generic Type-Parameter is a only a single capital letter, which is the usual Java convention. In jgraph-core there are 19 violation against those rules. For example in `BaseIntrusiveEdgesSpecifics` where IE is used as a parameter-name. In this case `I` would also be suitable in my opinion.
But for Methods an second integer is probably necessary for TypeParameterNames, otherwise the names would likely become odd.

What do you think about the other rules? Maybe it would be best to create PR's to apply/fix them in groups of similar rules so the effect can be discussed.

**Comment by jsichi** [comment]
I don't think there's any good reason to restrict type parameter names to only one letter.

**Comment by jsichi** [comment]
For the remaining rules, yes, please go ahead and create PR's for fixing/enforcing them in groups.

**Comment by HannesWell** [comment]
> I don't think there's any good reason to restrict type parameter names to only one letter.

The only reason I know is the convention and that people are used to it. Furthermore it would be really hard to find suitable fixes for the violations against the different `TypeParameterName` rules. Nevertheless an adjusted set of rules were one capital letter and one optional following capital letter or one digit are allowed would be reasonable for me. The only violation against such rules is the `AbstractGraphIterator` that has a type-parameter name `localE`. Do you think such adjusted rules are suitable or should the be discarded totally?

> For the remaining rules, yes, please go ahead and create PR's for fixing/enforcing them in groups.

The only rules left besides the TypeNameParamter-rules is the `CatchParameterName`-rule which has only a few violations at the moment. I will submit a PR.


As an addition to #1068:
It would also be possible to enforce the specified Unicode character-set for comments, String literals and text blocks. In jgrapht-core there would be 24 violations in `VF2GraphIsomorphismState` and `VF2SubgraphIsomorphismState` because it uses mathematical set-symbols in Exception message. I find this reasonable and therefore to not apply this rule.

**Comment by jsichi** [comment]
TypeParameterName:  your adjusted rule seems good to me.

Unicode character-set restrictions: no need to add any further restrictions.

**Comment by HannesWell** [comment]
> TypeParameterName: your adjusted rule seems good to me.
> 
> Unicode character-set restrictions: no need to add any further restrictions.

Great. So with the PR I just submitted to me it looks like all discussed checkstyle rules are now applied to JGraphT.

Do you have more rules you want to apply?

Nevertheless I would like to keep this issue open until I've evaluated the additional analyzer tools I suggested in issue #1056.
For example PMD has a `MissingOverride` rule. Maybe this works better to enforce @Override annotations were recommended, which was also suggested but what Checkstyle cannot do properly.

**Comment by jsichi** [comment]
For now, I don't have any more to suggest.
